src/Chapter3.hs

```diff
@@ -971,6 +1092,42 @@ implement the following functions:
 🕯 HINT: to implement this task, derive some standard typeclasses
 -}
 
+data Weekday = Monday | Tuesday | Wednesday | Thursday | Friday | Saturday | Sunday
+  deriving (Show, Eq, Bounded, Enum)
+
+isWeekend :: Weekday -> Bool
+isWeekend day
+  | day == pred maxBound  = True
+  | day == maxBound       = True
+  | otherwise             = False
+
+{-
+-- I would like to be able to express the cycling property of the function nextDay in terms of its minBound and maxBound
+-- I've had no luck so far...
+nextDay :: Weekday -> Weekday
+nextDay maxBound = minBound
+nextDay x = succ x
+-}
+
+-- I'm not quite happy with this either because I would have loved to use succ. Oh well w/e.
+nextDay :: Weekday -> Weekday
+nextDay x = cycle [minBound .. maxBound] !! (fromEnum x + 1)
```

Comment:
Your commented solution doesn't work because during pattern matching you don't compare the input variable with `maxBound`. You just give the name `maxBound` to your variable, and the variable name is a pattern that always succeeds. This is possible due to _name shadoing_. In Haskell, you can override global names with local names. Fortunately, GHC produces a warning in such situations, so you can fix them.

Instead of pattern matching on `maxBound`, compare your input with `maxBound`. And you will be able to use `succ` in this case:

```haskell
nextDay day
    | day == maxBound = minBound
    | otherwise = ...
```